base/error.jl

```diff
@@ -50,16 +50,20 @@ macro assert(ex, msgs...)
     :($(esc(ex)) ? $(nothing) : throw(Main.Base.AssertionError($msg)))
 end
 
+## fatal errors ##
 
 isfatal(error) = false
-if isdefined(Main, :Base)
 isfatal(::StackOverflowError) = true
 isfatal(::OutOfMemoryError) = true
 isfatal(::UndefVarError) = true
-end
 
-set_fatal_eh() = ccall(:jl_set_fatal_eh, Void, ())
-rethrow_fatal() = ccall(:jl_rethrow_fatal, Void, ())
+enable_catch_fatal() = ccall(:jl_enable_catch_fatal, Void, ())
+disable_catch_fatal() = ccall(:jl_disable_catch_fatal, Void, ())
+if isdefined(Main, :Base)
+rethrow_if_fatal(error) = isfatal(error) && ccall(:jl_rethrow_fatal, Void, ())
+else
+rethrow_if_fatal(error) = nothing
+end
```

Comment:
Analysis:

[The `try/catch` in `pure_eval_call`](https://github.com/JuliaLang/julia/blob/df6059fa60c6f7610442c4dcb77705c541ce283a/base/inference.jl#L1011) ended up calling to `promote_to_supertype`, which was [calling `error("no promotion exists for ", Float16, " and ", Int64)`](https://github.com/JuliaLang/julia/blob/df6059fa60c6f7610442c4dcb77705c541ce283a/base/promotion.jl#L170). However, `error` was defined as `throw(Main.Base.ErrorException(Main.Base.string(s...)))`, and `Base` does not exist when building `coreimg.jl`,  so the result was `UndefVarError(var=:Base)`.

Fix:

Define a simple version of `error` that works without `Base`:
```julia
    type CoreErrorException <: Exception
        msg
    end
    error(s...) = throw(CoreErrorException(s))
```